src/providers/googleProvider.ts

```diff
@@ -1,59 +1,119 @@
-    return this.getUrl(this.searchUrl, params);
+  constructor(geoCoder: google.maps.Geocoder) {
+    super();
+    this._checkGeoCoder(geoCoder);
+    this.geoCoder = geoCoder;
   }
 
-  parse(result: ParseArgument<RequestResult>): SearchResult<RawResult>[] {
-    return result.data.results.map((r) => ({
-      x: r.geometry.location.lng,
-      y: r.geometry.location.lat,
-      label: r.formatted_address,
-      bounds: [
-        [r.geometry.viewport.southwest.lat, r.geometry.viewport.southwest.lng], // s, w
-        [r.geometry.viewport.northeast.lat, r.geometry.viewport.northeast.lng], // n, e
-      ],
-      raw: r,
-    }));
+  static async loadGeocoder(
+    options: ProviderOptions = {},
+  ): Promise<google.maps.Geocoder> {
+    return new Loader({
+      // The toString() is necessary to address a type mismatch -
+      // key can be string | number | boolean, but Google Maps
+      // expects it to be a string.
+      apiKey: options.params?.key?.toString() || '',
+      version: 'weekly',
+      // If a region is not specified, geocoding requests will bias toward the United States.
+      region: options.params?.region?.toString(),
+    })
+      .load()
+      .then((google) => {
+        return new google.maps.Geocoder();
+      });
+  }
+
+  endpoint({ query }: EndpointArgument): never {
+    throw new Error('Method not implemented.');
+  }
+
+  parse(
+    response: ParseArgument<RequestResult>,
+  ): SearchResult<google.maps.GeocoderResult>[] {
+    return response.data.results.map((r) => {
+      const { lat, lng } = r.geometry.location.toJSON();
+      const { east, north, south, west } = r.geometry.viewport.toJSON();
+
+      return {
+        x: lng,
+        y: lat,
+        label: r.formatted_address,
+        bounds: [
+          [south, west],
+          [north, east],
+        ],
+        raw: r,
+      };
+    });
+  }
+
+  _checkGeoCoder(geoCoder: google.maps.Geocoder) {
+    if (geoCoder && geoCoder.hasOwnProperty('geocode')) {
+      return;
+    }
+
+    throw new Error(
+      'You have to provide a google.maps.Geocoder object before instantiating this provider. Please consult the docs.',
+    );
+  }
+
+  async search(
+    options: SearchArgument,
+  ): Promise<SearchResult<google.maps.GeocoderResult>[]> {
+    const response = await this.geoCoder
+      .geocode(
+        {
+          // TODO (maybe?) - a user can specify a region when loading the Google Maps JS API,
+          // but a region can also be provided to individual requests. Is that something we'd
+          // want to support now, or perhaps in a future version?
+          address: options.query,
+        },
+        (
+          response: google.maps.GeocoderResult[] | null,
+          status: google.maps.GeocoderStatus,
+        ) => {
+          return { results: response ? response : [] };
+        },
+      )
+      .catch((e: GeocodeError) => {
+        console.error(`${e.code}: ${e.message}`);
```

Comment:
When I tried it out in my app, any search query yielding zero results would fall into the catch block.
In other words if you type gibberish, your console turns red. I think we should either remove this or figure out why no results is considered an error. wdyt?

Reply:
Totally agree - no results shouldn't be an error. I think the way you handled it here is perfect -

```
      .catch((e: GeocodeError) => {
        if (e.code !== 'ZERO_RESULTS') {
          console.error(`${e.code}: ${e.message}`);
        }
        return { results: [] };
      });
```

I hadn't gotten far enough with my work on this to see that happen yet! 😅